File: crates/app/src/community.rs
//! The Community tab's commands: the kendex.ai directory (cached on disk,
//! honest about staleness) and skills.sh search — thin shells over core's
//! registry, like every other command here.

use kendex_core::registry::CurlFetch;
use kendex_core::registry::skillssh::{self, SkillsShHit};
use kendex_core::registry::view::{self, DirectoryView};

use crate::scopes::env;

/// The directory as the tab shows it. `refresh` forces a revalidation;
/// otherwise the cached list is served within its TTL.
#[tauri::command(async)]
#[specta::specta]
pub fn community_directory(refresh: bool) -> Result<DirectoryView, String> {
    let env = env()?;
    view::directory(&env, &CurlFetch, refresh).map_err(|e| e.to_string())
}

/// Search skills.sh directly — public API, no account, only skills.sh
/// sees the query.
#[tauri::command(async)]
#[specta::specta]
pub fn community_skillssh_search(query: String) -> Result<Vec<SkillsShHit>, String> {
    if community_skillssh_enabled()? {
        skillssh::search(&CurlFetch, &query).map_err(|e| e.to_string())
    } else {
        Err("skills.sh is switched off".into())
    }
}

/// Trending / Hot / Top through the kendex.ai proxy. An error here means
/// "no proxy" — the chips hide, search stays.
#[tauri::command(async)]
#[specta::specta]
pub fn community_skillssh_leaderboard(view: String) -> Result<Vec<SkillsShHit>, String> {
    let Some(view) = skillssh::LeaderboardView::parse(&view) else {
        return Err(format!("'{view}' is not a leaderboard view"));
    };
    skillssh::leaderboard(&CurlFetch, view).map_err(|e| e.to_string())
}

/// Whether the skills.sh surface is on at all — the tab hides the
/// sub-tab when it is not, rather than showing a dead search box.
#[tauri::command(async)]
#[specta::specta]
pub fn community_skillssh_available() -> Result<bool, String> {
    community_skillssh_enabled()
}

fn community_skillssh_enabled() -> Result<bool, String> {
    Ok(skillssh::enabled())
}

File: ui/src/stores/community.ts
// The Community tab's state: the kendex.ai directory (served from the
// app's on-disk cache, honest about staleness) and skills.sh search plus
// its proxied leaderboard.
import { create } from "zustand";
import { commands, type DirectoryView, type SkillsShHit } from "@/bindings";
import { readOrder } from "@/lib/read-state";

export type SkillsShMode = "search" | "all-time" | "trending" | "hot";

interface CommunityState {
  directory: DirectoryView | null;
  loading: boolean;
  /** Set only when there is nothing to show at all — a stale list renders
   * with its "as of" line instead. */
  error: string | null;

  skillsshAvailable: boolean;
  skillsshHits: SkillsShHit[] | null;
  skillsshMode: SkillsShMode;
  /** The leaderboard needs the kendex.ai proxy; when it is not there the
   * chips hide and search carries the whole sub-tab. */
  skillsshChips: boolean;
  skillsshSearching: boolean;
  skillsshError: string | null;

  load: (refresh: boolean) => Promise<void>;
  searchSkillssh: (query: string) => Promise<void>;
  loadLeaderboard: (view: Exclude<SkillsShMode, "search">) => Promise<void>;
}

/** Stale results never land on a newer query: the search box asks on every
 *  keystroke and a chip asks beside it, so several are routinely out. */
const order = readOrder();

export const useCommunityStore = create<CommunityState>((set) => ({
  directory: null,
  loading: false,
  error: null,
  skillsshAvailable: true,
  skillsshHits: null,
  skillsshMode: "search",
  skillsshChips: true,
  skillsshSearching: false,
  skillsshError: null,

  load: async (refresh) => {
    set({ loading: true });
    try {
      const [view, available] = await Promise.all([
        commands.communityDirectory(refresh),
        commands.communitySkillsshAvailable(),
      ]);
      if (view.status === "ok") {
        set({ directory: view.data, error: null });
      } else {
        set({ error: view.error });
      }
      set({
        skillsshAvailable: available.status === "ok" ? available.data : false,
      });
    } finally {
      set({ loading: false });
    }
  },

  searchSkillssh: async (query) => {
    const ticket = order.begin();
    if (!query.trim()) {
      set({
        skillsshHits: null,
        skillsshError: null,
        skillsshSearching: false,
        skillsshMode: "search",
      });
      return;
    }
    set({ skillsshSearching: true, skillsshMode: "search" });
    const response = await commands.communitySkillsshSearch(query);
    if (!order.lands(ticket)) return;
    if (response.status === "ok") {
      set({
        skillsshHits: response.data,
        skillsshError: null,
        skillsshSearching: false,
      });
    } else {
      set({ skillsshError: response.error, skillsshSearching: false });
    }
  },

  loadLeaderboard: async (view) => {
    const ticket = order.begin();
    set({ skillsshSearching: true });
    const response = await commands.communitySkillsshLeaderboard(view);
    if (!order.lands(ticket)) return;
    if (response.status === "ok") {
      set({
        skillsshHits: response.data,
        skillsshMode: view,
        skillsshError: null,
        skillsshSearching: false,
      });
    } else {
      // No proxy (or it is down): the chips disappear rather than sit
      // broken; whatever the list showed stays.
      set({ skillsshChips: false, skillsshSearching: false });
    }
  },
}));
